### ISSUE TITLE
[Common Lisp] Implement new Concept Exercise: higher-order-functions

### ISSUE THREAD

**Issue by TheLostLambda** [exercism/v3#1710]
This issue describes how to implement the `higher-order-functions` concept exercise for the Common Lisp track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Learning objectives

- Understand `funcall` and `apply`
- Be comfortable working with "mapping" functions like `mapcar` or `map`
- Be comfortable working with "filter" functions like `remove-if` and `remove-if-not`
- Be comfortable working with "reduce" functions like `reduce`
- Be able to use the "mapping" functions with several lists, performing a "zip"
- Know how to use `:key` functions to "combine" map and reduce operations
- Know to use `#'` instead of `'` for type-checking and lisp-2 disambiguation
- Be exposed to other higher-order functions like `sort`, `every`, `some`, `count-if`, and `find-if`

## Out of scope

- "Mapping" functions other than `mapcar` and `map`
- Anonymous functions and `lambda`
- Local functions with `flet` or `labels`
- All of the keyword arguments of these functions
- Hash-tables
- Macros like `loop` and the `do` family

## Concepts

- `higher-order-functions`

## Prerequisites

- `basic-lists`
- `strings`
- `function-definition`
- `truthy-and-falsy`
- `symbols`

## Resources to refer to

- [Functions As Data in Practical Common Lisp][http://www.gigamonkeys.com/book/functions.html]

### Hints

- [RIP Tutorial][https://riptutorial.com/common-lisp/topic/1259/functions-as-first-class-values]

### After

- None yet, any ideas?

## Representer

Nothing yet

## Analyzer

- Enforce the use of "mapping" functions instead of `loop` or `do`
- Similar checking for "reduce" and "filter" functions
- Ensure that `#'` is being used and not `'`, though both are likely to pass tests

## Implementing

Perhaps start with some data and do a map-filter-reduce to calculate something from it. You can get some other statistics using `count-if` and `find-if` and `sort`ing is easy to show off.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.
